plugins/modules/gpg_keypair.py

```diff
@@ -0,0 +1,507 @@
+    )
+
+    rc, stdout, stderr = module.run_command(
+        [
+            '--dry-run' if check_mode else '',
+            '--batch',
+            '--gen-key',
+        ],
+        data=parameters,
+        executable=get_bin_path('gpg'),
+        check_rc=True
+    )
+
+    fingerprint = re.search(r'([0-9A-Z]+)\.rev', stderr).group(1)
+
+    for subkey in params['subkeys']:
+        if subkey['subkey_type'] in ['RSA', 'DSA', 'ELG']:
+            algo = '{0}'.format(subkey['subkey_type'].lower())
+            if subkey['subkey_length']:
+                algo += str(subkey['subkey_length'])
+        elif subkey['subkey_curve']:
+            algo = subkey['subkey_curve']
+        else:
+            algo = 'default'
+
+        module.run_command(
+            [
+                '--dry-run' if check_mode else '',
+                '--batch',
+                '--passphrase {0}'.format(params['passphrase']) if params['passphrase'] else ''
+                '--quick-add-key',
+                fingerprint,
+                algo,
+                ' '.join(subkey['subkey_usage']),
+                params['expire_date'] if params['expire_date'] else ''
+            ],
+            executable=get_bin_path('gpg')
+        )
+
+    return dict(changed=True, fingerprints=[fingerprint])
+
+
+def run_module(module, params, check_mode=False):
+    if params['expire_date']:
+        if not (params['expire_date'].isnumeric() or params['expire_date'][:-1].isnumeric()):
+            module.fail_json('Invalid format for expire date')
+
+    validate_key(module, params['key_type'], params['key_length'], params['key_curve'], params['key_usage'])
+    for i, subkey in enumerate(params['subkeys']):
+        validate_key(module, subkey['subkey_type'], subkey['subkey_length'], subkey['subkey_curve'], subkey['subkey_usage'], 'subkey #{0}'.format(i + 1))
+
+    uid = ''
+    if params['name']:
+        uid += '{0} '.format(params['name'])
+    if params['comment']:
+        uid += '({0}) '.format(params['comment'])
+    if params['email']:
+        uid += '<{0}>'.format(params['email'])
+    if uid:
+        uid = '"{0}"'.format(uid.strip())
+
+    rc, stdout, stderr = module.run_command(
+        ['--list-secret-keys', uid] + params['fingerprints'],
+        executable=get_bin_path('gpg')
+    )
+
+    fingerprints = list(set([line.strip() for line in stdout.splitlines() if line.strip().isalnum()]))
+    matching_keys = []
+    for fingerprint in fingerprints:
+        rc, stdout, stderr = module.run_command(
+            ['--list-secret-keys', '--with-colons', fingerprint],
+            executable=get_bin_path('gpg'),
+            check_rc=True
+        )
+
+        subkey_count = 0
+        uid_present = not bool(uid)
+        for line in stdout.splitlines():
+            if line[:3] == 'sec':
+                primary_key = re.search(r'.+:([0-9]+):([0-9]+):[0-9A-Z]+:[0-9]+:[0-9]+::u:+([a-z]+)[A-Z]+:+\+:+([0-9a-zA-Z]+)::0:', line)
```

Comment:
Why not simply split by `:`, and then access the fields?

Reply:
Much simpiler than I was thinking haha. Seems like a good change.